spongefish/src/pattern/interaction_pattern.rs

```diff
@@ -0,0 +1,190 @@
+use core::fmt::Display;
+
+use thiserror::Error;
+
+use super::{interaction::Hierarchy, Interaction, Kind};
+
+/// Abstract transcript containing prover-verifier interactions
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Debug, Default)]
+pub struct InteractionPattern {
+    interactions: Vec<Interaction>,
+}
+
+/// Errors when validating a transcript.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash, Error)]
+pub enum TranscriptError {
+    #[error("Missing Begin for {end} at {position}")]
+    MissingBegin { position: usize, end: Interaction },
+    #[error(
+        "Invalid kind {interaction} at {interaction_position} for {begin} at {begin_position}"
+    )]
+    InvalidKind {
+        begin_position: usize,
+        begin: Interaction,
+        interaction_position: usize,
+        interaction: Interaction,
+    },
+    #[error("Mismatch {begin} at {begin_position} for {end} at {end_position}")]
+    MismatchedBeginEnd {
+        begin_position: usize,
+        begin: Interaction,
+        end_position: usize,
+        end: Interaction,
+    },
+    #[error("Missing End for {begin} at {position}")]
+    MissingEnd { position: usize, begin: Interaction },
+}
+
+impl InteractionPattern {
+    pub fn new(interactions: Vec<Interaction>) -> Result<Self, TranscriptError> {
+        let result = Self { interactions };
+        result.validate()?;
+        Ok(result)
+    }
+
+    #[must_use]
+    #[allow(clippy::missing_const_for_fn)] // False positive
+    pub fn interactions(&self) -> &[Interaction] {
+        &self.interactions
+    }
+
+    /// Generate a unique identifier for the protocol.
+    ///
+    /// It is created by taking the SHA3 hash of a stable unambiguous
+    /// string representation of the transcript interactions.
+    // TODO: A more neutral implementation would use ASN.1 DER.
+    #[must_use]
+    pub fn domain_separator(&self) -> [u8; 32] {
+        use sha3::{Digest, Sha3_256};
+        let mut hasher = Sha3_256::new();
+        // Use Display in `alternate` mode for stable unambiguous representation.
+        hasher.update(format!("{self:#}").as_bytes());
+        let result = hasher.finalize();
+        result.into()
+    }
+
+    /// Validate the transcript.
+    ///
+    /// A valid transcript has:
+    ///
+    /// - Matching [`InteractionHierarchy::Begin`] and [`InteractionHierarchy::End`] interactions
+    ///   creating a nested hierarchy.
+    /// - Nested interactions are the same [`InteractionKind`] as the last [`InteractionHierarchy::Begin`] interaction, except for [`InteractionKind::Protocol`] which can contain any [`InteractionKind`].
+    fn validate(&self) -> Result<(), TranscriptError> {
```

Comment:
Should we fuzz this just to be safe?

Reply:
That would be great, but I think it's hard to come up with a specification that simpler than the function itself.

A fuzzer is basically a poor-mans formal verification, where you test an implementation against a specification on random inputs. But that only helps if the specification is easier to verify than the implementation itself, which is hard to achieve here.